[ready]Workaround for Out of resources byond bug

MrsTonedOne opened the pull request:
Fixes #9503 - Out of resources runtime (dubbed "the mc crashing" for whatever fucking reason.)
Fixes #7410 (same as above)

~~Simply put, running locate in locs causes the copy of the list byond creates in locate to never get deallocated.~~ actually, it turns out that it's any accessing of atom/movable.locs.

If you later run a "thing in things" style for loop after hitting some limit (2^24) on the total number of lists, an out of resources runtime will spew, everything will break, and attempts to access lists will randomly fail and cause that list to get corrupted.

Test Case: http://pastebin.com/1kiXQb6s

~~I haven't actually tested this, i'm going to bed. test before merging if you decide to quick merge this.~~

~~I tested it, shit is runtiming about something being null when we just checked it the line before. most likely ANOTHER byond bug.~~

~~do not merge quite yet.~~

Ready.

MrsTonedOne committed:
Workaround for Out of resources byond bug

MrsTonedOne committed:
fixes compile error (hopefully)

Jalleo commented:
Controversial since its not tested?

Dorsisdwarf commented:
That's not what the Controversial tag means. If a bunch of people came into the PR and started going on about how it would ruin the game, that's what the tag is for.

Miauw62 commented:
this will ruin the game though, common crashes keep rounds from dragging on too long and thus turning us into bay, causing the decay of tgstation culture and many more special snowflakes and the end of the world.

Incoming5643 commented:
good to know we can only have a septillion instanced lists

RemieRichards commented:
```locs``` what a hitler.

:+1: :heavy_multiplication_x: :100:  Great work MSO.

GunHog commented:
Maybe Sticky can finally run his 8 hour rounds without them crashing!

MrStonedOne commented:
tp://www.byond.com/forum/?post=1852790 "Lummox JR resolved issue with message: Reading a movable atom's locs list caused a memory leak."

we have quite a few more locs in the code. and no idea when this beta build is gonna come out. fun...

RemieRichards commented on code/__HELPERS/unsorted.dm:
```diff
@@ -1074,7 +1074,14 @@ Turf and target are seperate in case you want to teleport some distance from a t
 	if(istype(AM))
-		return locate(/turf) in AM.locs
+		var/list/atom/checkedatoms = list() //prevent recursion from badmins being dumbasses
+		var/atom/A = AM
```
@MrStonedOne can't you skip one step by changing this to ```var/atom/A = AM.loc```
Given that in this case AM will always be a /atom/movable not a turf, thus allowing us to skip the first iteration of the ```while(!isturf(A))``` loop.

optimumtact commented:
Do not wait

MrsTonedOne committed:
Minor workaround for the runtime in get_turf on world new.

Adding this second check for turf_candidate being null FOR SOME FUCKING REASON is needed.

MrsTonedOne committed:
Utterly guts the use of .locs from the game.

They

This means that multi-tile things may not handle correctly, but the only thing that used it was atmos for double wide doors, and they are used in one location that is unlikely to have atmos changes

see: http://www.byond.com/forum/?post=1852790 for justifcation. (tl;dr: accessing .locs causes memory leaks)

MrStonedOne commented:
@RemieRichards @Miauw62 @Jordie0608 @Cheridan this are ready

@Stickymayhem highlighting you so you know when this is merged.

Menshin commented:
:+1: :+1: :+1:
